leader/election.go

```diff
@@ -1,139 +1,131 @@
+		return "", fmt.Errorf("failed to read namespace file: %w", err)
 	}
 
-}
-
-func notify(isLeader bool) {
-	for _, fn := range watchers {
-		fn(isLeader)
+	ns := strings.TrimSpace(string(data))
+	if ns == "" {
+		return "", errors.New("namespace was neither found in the env nor in the service account path")
 	}
 
+	return ns, nil
 }
 
-func IsLeader(ctx context.Context, leaseName string) (bool, error) {
-	leases := ctx.Kubernetes().CoordinationV1().Leases(ctx.GetNamespace())
-
-	lease, err := leases.Get(ctx, leaseName, metav1.GetOptions{})
+func init() {
+	var err error
+	hostname, err = os.Hostname()
 	if err != nil {
-		return false, err
+		log.Fatalf("failed to get hostname: %v", err)
 	}
 
-	if *lease.Spec.HolderIdentity == identity {
-		return true, nil
+	if n, err := getPodNamespace(); err == nil {
+		podNamespace = n
 	}
 
-	return false, nil
+	service = strings.Split(hostname, "-")[0]
 }
 
-func createOrUpdateLease(ctx context.Context, leaseName string, attempt int) (bool, string, error) {
-	if attempt > 0 {
-		time.Sleep(time.Duration(rand.Intn(5000)) * time.Millisecond)
-	}
-	if attempt >= ctx.Properties().Int("leader.lease.attempts", 3) {
-		return false, "", fmt.Errorf("failed to acquire lease %s after %d attempts", leaseName, attempt)
+func Register(
+	ctx context.Context,
+	namespace string,
+	onLead func(ctx gocontext.Context),
+	onStoppedLead func(),
+	onNewLeader func(identity string),
+) {
+	if namespace == "" {
+		namespace = podNamespace
 	}
-	now := metav1.MicroTime{Time: time.Now()}
-	leases := ctx.Kubernetes().CoordinationV1().Leases(ctx.GetNamespace())
-	leaseDuration := ctx.Properties().Duration("leader.lease.duration", 10*time.Minute)
-	lease, err := leases.Get(ctx, leaseName, metav1.GetOptions{})
-	if err != nil {
-		return false, "", err
+
+	ctx = ctx.WithNamespace(namespace)
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      service,
+			Namespace: namespace,
+		},
+		Client: ctx.Kubernetes().CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: hostname,
+		},
 	}
-	if lease == nil {
-		ctx.Infof("Acquiring lease %s", leaseName)
-		lease = &v1.Lease{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      leaseName,
-				Namespace: ctx.GetNamespace(),
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: true,
+		LeaseDuration:   ctx.Properties().Duration("leader.lease.duration", 30*time.Second),
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(leadCtx gocontext.Context) {
+				updateLeaderLabel(ctx, true)
+				onLead(leadCtx)
 			},
-			Spec: v1.LeaseSpec{
-				HolderIdentity:       &identity,
-				LeaseDurationSeconds: lo.ToPtr(int32(leaseDuration.Seconds())),
-				AcquireTime:          &now,
-				RenewTime:            &now,
+			OnStoppedLeading: func() {
+				updateLeaderLabel(ctx, false)
```

Comment:
I don't think we can rely on this to add a label - if the pod is killed or partitions it will never run - The leader should first add its own label, and then remove the label from other others

Reply:
I think we should also stop and start all schedules here